src/napari_threedee/visualization/camera_spline.py

```diff
@@ -0,0 +1,250 @@
+        else:
+            self.spline_valid = True
+
+    @property
+    def image_layer(self) -> Image:
+        return self._image_layer
+
+    @image_layer.setter
+    def image_layer(self, layer: Image):
+        self._image_layer = layer
+        # self.spline_annotator_model.set_layers(layer)
+
+    def start_spline_annotation(self):
+        self.spline_annotator_model.enabled = True
+
+        # disable the key binding to switch to the next spline index
+        self.viewer.bind_key('n', None)
+
+    def stop_spline_annotation(self):
+        self.spline_annotator_model.enabled = False
+
+    def _on_enable(self):
+        self.spline_annotator_model.set_layers(self.image_layer)
+
+    def _on_disable(self):
+        self.spline_annotator_model.enabled = False
+        self.mode = CameraSplineMode("pan_zoom")
+
+    def set_layers(self, image_layer: napari.layers.Image):
+        self.image_layer = image_layer
+
+    def set_camera_position(self, spline_coordinate: float):
+        if self.image_layer is None:
+            # do not do anything if the image layer hasn't been set
+            return
+
+        spline_dict = self.spline_annotator_model.points_layer.metadata["splines"]
+
+        # only one spline
+        spline_object = spline_dict[0]
+        spline_point = spline_object._sample_backbone(u=[spline_coordinate])
+        self.viewer.camera.center = np.squeeze(spline_point)
+
+        view_direction = np.squeeze(spline_object._sample_backbone(u=[spline_coordinate], derivative=1))
+        view_direction_displayed = view_direction[list(self.viewer.dims.displayed)]
+
+        view_direction_transformed = tuple(self._transform_view_direction(view_direction_displayed))
+
+        self.viewer.camera.set_view_direction(
+            view_direction=view_direction_transformed, up_direction=self.up_direction
+        )
+
+    def _transform_view_direction(self, view_direction: np.ndarray) -> np.ndarray:
+        """Transform the view direction along the spline to the view set by the user.
+
+        The transformation applied is in self.view_direction_transformation.
+
+        Parameters
+        ----------
+        view_direction : np.ndarray
+            The view direction in displayed data coordinates.
+            Should be a (3,) array.
+
+        Returns
+        -------
+        transformed_view_direction : np.ndarray
+            The view direction transformed by self.view_direction_transformation
+        """
+        return self.view_direction_transformation.dot(view_direction)
+
+    def start_spline_exploration(self):
+        """Setup up the exploration mode.
+
+        This is called when mode is switched to exploration mode.
+        """
+        self.set_camera_position(self.current_spline_coordinate)
+
+    def stop_spline_exploration(self):
+        """Clean up the exploration mode.
+
+        This is called when mode is switched from exploration mode.
+        """
+        pass
```

Comment:
could the lack of an implementation here be the source of the mode switching bug?

Reply:
I figured it out and there isn't a bug. It's just that I had perspective on and the mouse callback for moving the plane no longer worked since `get_value` doesn't work with perspective 😆